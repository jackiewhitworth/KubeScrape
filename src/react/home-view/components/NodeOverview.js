/**
 * ************************************
 *
 * @module  NodeOverview.js
 * @description component that renders information about a node
 *
 * ************************************
 */

import React, { useState, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { useSelector, useDispatch } from 'react-redux';
import GaugeChart from 'react-gauge-chart';
import { Container, Grid, Paper, Typography } from '@mui/material';
import { styled } from '@mui/material/styles';
import * as actions from '../../../redux/actions/actions';

// fetch requests to the Prometheus server are stored as functions in PrometheusAPI/node-promql-requests.js
import * as nodePromql from '../../../PrometheusAPI/node-promql-requests';

// create a functional component
const NodeOverview = props => {
  // useSelector allows you to extract data from the Redux store state, using a selector function
  // this function accesses the state from the nodeReducer by subscribing to the store through sseSelector
  const { nodeCpuUsage, nodeMemoryUsage, pods, nodePodCapacity } = useSelector(state => state.node);
  const [nodeNetworkUtilization, setNodeNetworkUtilization] = useState(0);
  const [nodeNetworkErrors, setNodeNetworkErrors] = useState(0);

  // the useDispatch hook returns a reference to the dispatch function from the Redux store.
  // dispatch can now be used to dispatch actions as needed
  const dispatch = useDispatch();

  const history = useHistory();

  // function to fetch prometheus data and store using redux
  const fetchDataToStore = async () => {
    const nodeCpuUsagePercentage = await nodePromql.fetchCpuUsage();
    const nodeMemoryUsagePercentage = await nodePromql.fetchMemoryUsage(props.nodeName);
    const pods = await nodePromql.fetchNodePods(props.nodeName);
    const nodePodCapacity = await nodePromql.fetchPodCapacity();
    const currentNetworkUtilization = await nodePromql.fetchNetworkUtilization();
    const currentNetworkErrors = await nodePromql.fetchNetworkErrors();

    dispatch(actions.setCpuUsage(nodeCpuUsagePercentage));
    dispatch(actions.setMemoryUsage(nodeMemoryUsagePercentage));
    dispatch(actions.setNodePods(pods));
    dispatch(actions.setPodCapacity(nodePodCapacity));
    setNodeNetworkUtilization(currentNetworkUtilization);
    setNodeNetworkErrors(currentNetworkErrors);
  };

  // fetch data, then fetch again in 30 seconds
  // when component unmounts, cancel setInterval in a cleanup function
  useEffect(() => {
    fetchDataToStore();
    const interval = setInterval(() => fetchDataToStore(), 30000);
    return () => clearInterval(interval);
  }, []);

  // Styles
  const PREFIX = 'NodeOverview';
  const classes = {
    flex: `${PREFIX}-flex`,
    graphItem: `${PREFIX}-graphItem`,
    metricsItem: `${PREFIX}-metrixItem`,
  };
  const StyledContainer = styled(Container)(({ theme }) => ({
    ':hover': {
      filter: 'brightness(150%)',
      cursor: 'pointer',
    },
  }));
  const GridItem = styled(Grid)(({ theme }) => ({
    [`&.${classes.flex}`]: {
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'flex-start',
      alignItems: 'center',
    },
    [`&.${classes.graphItem}`]: {
      marginTop: '-15px',
    },
    [`&.${classes.metricsItem}`]: {
      marginTop: '25px',
    },
  }));

  // function to handle node click
  const goToNode = nodeName => {
    history.push({
      pathname: '/node',
      nodeName,
    });
  };

  // function to render a gauge
  const renderGauge = (title, value) => (
    <>
      <h6>{title}</h6>
      <GaugeChart
        id="gauge-chart"
        nrOfLevels={3}
        colors={['#29648A', '#F8E9A1', '#F76C6C']}
        arcWidth={0.3}
        arcPadding={0}
        cornerRadius={0}
        percent={value}
        textColor="#FFF"
        needleColor="#FFF"
        animate={false}
      />
    </>
  );

  return (
    <StyledContainer
      maxWidth="xs"
      onClick={() => goToNode(props.nodeName)}
    >
      <Paper elevation={3}>
        <Typography
          variant="h6"
          component="div"
          align="center"
          sx={{
            paddingTop: '20px',
          }}
        >
          Node: {props.nodeName}
        </Typography>
        <Grid container justifyContent="center">
          <GridItem item sm={6} lg={3} className={`${classes.flex} ${classes.metricsItem}`}>
            <span>{pods.length}</span>
            <h6>Active Pods</h6>
          </GridItem>
          <GridItem item sm={6} lg={3} className={`${classes.flex} ${classes.metricsItem}`}>
            <span>{nodePodCapacity - pods.length}</span>
            <h6>Available Pods</h6>
          </GridItem>
          <GridItem item sm={6} lg={3} className={`${classes.flex} ${classes.metricsItem}`}>
            <span>{nodeNetworkUtilization} kb/s</span>
            <h6>Network Utilization</h6>
          </GridItem>
          <GridItem item sm={6} lg={3} className={`${classes.flex} ${classes.metricsItem}`}>
            <span>{nodeNetworkErrors}</span>
            <h6>Errors</h6>
          </GridItem>
          <GridItem item lg={6} className={`${classes.flex} ${classes.graphItem}`}>
            {renderGauge('Node CPU Usage', nodeCpuUsage / 100)}
          </GridItem>
          <GridItem item lg={6} className={`${classes.flex} ${classes.graphItem}`}>
            {renderGauge('Node Memory Usage', nodeMemoryUsage / 100)}
          </GridItem>
        </Grid>
      </Paper>
    </StyledContainer>
  );
};

export default NodeOverview;
