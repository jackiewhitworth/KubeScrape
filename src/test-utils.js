import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from './redux/store';


export const renderWithRedux = (
  component,
  reduxStore = store
) => render(<Provider store={reduxStore}>{component}</Provider>);

export const fakeClusterCpuUsage = {
  'status': 'success',
  'data': {
    'resultType': 'vector',
    'result': [
      {
        'metric': {
          'instance': '10.1.2.240:9100'
        },
        'value': [
          1639439695.158,
          '6.151261565921784'
        ]
      }
    ]
  }
};

