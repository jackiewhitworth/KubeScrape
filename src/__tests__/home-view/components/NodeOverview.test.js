/* eslint-disable no-undef */
import React from 'react';
import NodeOverview from '../../../react/home-view/components/NodeOverview';
import { renderWithRedux } from '../../../test-utils';


describe('NodeOverview', () => {

  it('renders and matches the snapshot', () => {
    const { container } = renderWithRedux(<NodeOverview />);
    expect(container).toMatchSnapshot();
  });

});
