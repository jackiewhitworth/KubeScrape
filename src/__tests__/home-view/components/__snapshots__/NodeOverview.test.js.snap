// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeOverview renders and matches the snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthXs css-3subwl-MuiContainer-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation3 css-3su884-MuiPaper-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h6 MuiTypography-alignCenter css-hb1txa-MuiTypography-root"
      >
        Node: 
      </div>
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-undefined css-1lym95h-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6 MuiGrid-grid-lg-3 NodeOverview-flex NodeOverview-metrixItem css-1yhk66g-MuiGrid-root"
        >
          <span>
            0
          </span>
          <h6>
            Active Pods
          </h6>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6 MuiGrid-grid-lg-3 NodeOverview-flex NodeOverview-metrixItem css-1yhk66g-MuiGrid-root"
        >
          <span>
            0
          </span>
          <h6>
            Available Pods
          </h6>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6 MuiGrid-grid-lg-3 NodeOverview-flex NodeOverview-metrixItem css-1yhk66g-MuiGrid-root"
        >
          <span>
            0
             kb/s
          </span>
          <h6>
            Network Utilization
          </h6>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-sm-6 MuiGrid-grid-lg-3 NodeOverview-flex NodeOverview-metrixItem css-1yhk66g-MuiGrid-root"
        >
          <span>
            0
          </span>
          <h6>
            Errors
          </h6>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-6 NodeOverview-flex NodeOverview-graphItem css-7jaoyt-MuiGrid-root"
        >
          <h6>
            Node CPU Usage
          </h6>
          <div
            id="gauge-chart"
            style="width: 100%;"
          >
            <svg
              height="0"
              width="0"
            >
              <g
                transform="translate(0, 0)"
              >
                <g
                  class="doughnut"
                  transform="translate(0, 0)"
                >
                  <g
                    class="arc"
                  >
                    <path
                      d="M0,0Z"
                      style="fill: #29648A;"
                    />
                  </g>
                  <g
                    class="arc"
                  >
                    <path
                      d="M0,0Z"
                      style="fill: #F8E9A1;"
                    />
                  </g>
                  <g
                    class="arc"
                  >
                    <path
                      d="M0,0Z"
                      style="fill: #F76C6C;"
                    />
                  </g>
                </g>
                <g
                  class="needle"
                  transform="translate(0, 0)"
                >
                  <path
                    d="M 0 0 L 0 0 L 0 0"
                    fill="#FFF"
                  />
                  <circle
                    cx="0"
                    cy="0"
                    fill="#464A4F"
                    r="0"
                  />
                </g>
                <g
                  class="text-group"
                  transform="translate(0, 20)"
                >
                  <text
                    style="font-size: 0px; fill: #FFF; text-anchor: middle;"
                  >
                    0%
                  </text>
                </g>
              </g>
            </svg>
          </div>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-lg-6 NodeOverview-flex NodeOverview-graphItem css-7jaoyt-MuiGrid-root"
        >
          <h6>
            Node Memory Usage
          </h6>
          <div
            id="gauge-chart"
            style="width: 100%;"
          >
            <svg
              height="0"
              width="0"
            >
              <g
                transform="translate(0, 0)"
              >
                <g
                  class="doughnut"
                  transform="translate(0, 0)"
                >
                  <g
                    class="arc"
                  >
                    <path
                      d="M0,0Z"
                      style="fill: #29648A;"
                    />
                  </g>
                  <g
                    class="arc"
                  >
                    <path
                      d="M0,0Z"
                      style="fill: #F8E9A1;"
                    />
                  </g>
                  <g
                    class="arc"
                  >
                    <path
                      d="M0,0Z"
                      style="fill: #F76C6C;"
                    />
                  </g>
                </g>
                <g
                  class="needle"
                  transform="translate(0, 0)"
                >
                  <path
                    d="M 0 0 L 0 0 L 0 0"
                    fill="#FFF"
                  />
                  <circle
                    cx="0"
                    cy="0"
                    fill="#464A4F"
                    r="0"
                  />
                </g>
                <g
                  class="text-group"
                  transform="translate(0, 20)"
                >
                  <text
                    style="font-size: 0px; fill: #FFF; text-anchor: middle;"
                  >
                    0%
                  </text>
                </g>
              </g>
            </svg>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
